### PR TITLE
add uncertainty alongside entropy to validation output

### PR DESCRIPTION
**Pull request by MayMeta**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -595,18 +595,46 @@ def apply_dither(x: shaping.BT, dither: float):
 	else:
 		return x
 
+
 def entropy(log_probs, lengths = None, dim = 1, eps = 1e-9, sum = True, keepdim = False):
+	""" Calculate the entropy of probabilities, using log of probabilities, and applying the formula:
+			Entropy(p) = -SUM(pj * log pj).
```
Comment (vadimkantorov) [inline models.py:601]: I would say this is the classic entropy formula:
image (https://en.wikipedia.org/wiki/Entropy_(information_theory))

 it's definitely more "classic" than the formula in L603. So I think the formula in L603 is actually confusing

File: models.py [hunk 2]
```diff
@@ -595,18 +595,46 @@ def apply_dither(x: shaping.BT, dither: float):
...
+
+def weighted_mean_entropy(log_probs, lengths = None, dim = 1, eps = 1e-9):
+	""" Calculate the entropy of probabilities, using log of probabilities, and then take the weighted average of those
+		values, using (1 - silence_probability) as weight, where silence_probability is silence token (epsilon)
+		probability. This way, non-silent timeframes have larger weights, and silent timeframes have lower weights.
+		weighted_mean_entropy is a good measure for char-level model uncertainty, and correlates well with cer.
+	"""
+	prob = log_probs.exp()
+	e = -(prob * log_probs).sum(dim = dim)
+	silence_prob = prob[:, -1] if dim else prob[-1]
```
Comment (vadimkantorov) [inline models.py]: what is the meaning of `if dim` check? If `dim = 0`, the check will fall through to the `else` clause. Is it expected?

about the `-1`: If you wanted to select the silence idx, please create a function argument `eps_idx = -1` and add a `TODO` at the callsite to use tokenizer.eps_id

you may probably just replace this line with `silence_prob = prob.select(dim, eps_idx)`

If you want this function to work on batch-less 2d tensors, you can set default `dim = -2`

File: models.py [hunk 3]
```diff
@@ -595,18 +595,46 @@ def apply_dither(x: shaping.BT, dither: float):
...
+	weights = 1 - silence_prob
+	if lengths is not None:
+		mask = temporal_mask(e, lengths)
+		e *= mask
```
Comment (vadimkantorov) [inline models.py]: is this multiplication necessary? below you'd just multiply `e` by `weights` anyway. We also don't do this multiplication in the original `entropy(...)` function. if it's not necessary, please remove it

File: models.py (hunk 1 above)
Comment (MayMeta) [reply models.py:601]: Ok, I agree. Lines 602 & 603 removed.

File: models.py (hunk 2 above)
Comment (MayMeta) [reply models.py]: thx for the tip. I will use `prob.select(dim, eps_idx)`

File: models.py (hunk 3 above)
Comment (MayMeta) [reply models.py]: It is used in entropy function: line 605 `e = e * temporal_mask(e, lengths)`

but you are right, multiplying `weights` by `mask` is enough. fixed.

File: models.py [hunk 4]
```diff
@@ -595,18 +595,43 @@ def apply_dither(x: shaping.BT, dither: float):
...
-	) if sum else e
+	if not sum:
+		return e
+	elif lengths is None:
+		return e.mean(dim = -1)
+	else:
+		return e.sum(dim = -1) / (eps + lengths.type_as(log_probs))
+
+
+def weighted_mean_entropy(log_probs, lengths = None, dim = -2, eps = 1e-9, epsilon_ind = -1):
```
Comment (vadimkantorov) [inline models.py]: Please rename into `eps_id` to match existing naming in https://github.com/vadimkantorov/convasr/blob/9bab51ff907a95f39b107ac360c27e169237157a/text_tokenizers.py#L21

Comment (MayMeta) [reply models.py]: done

**Pull request merged**